### DISCUSSION TITLE
Questions about using HyperHDR Screen Capturing feature for an Interactive Art Installation

### DISCUSSION THREAD

**Discussion by mamanchekk** [awawa-dev/HyperHDR#775] [General]
Hello! I'm very new to HyperHDR, coding and addressable LEDs, so please be kind ;)

I'm looking to build an interactive audiovisual art installation using:

- a Raspberry Pi 4B
- a strip of 10m of WS2811 addressable LEDs
- two momentary switches

where the user has the option to press one of the two momentary switches to run two different low-resolution videos (with audio) that guide them through a meditation experience. The video file will control the look of the individual LEDs in a fashion similar (but way simpler) to this example:

[https://www.youtube.com/watch?v=VgcwcmokPeQ](https://www.youtube.com/watch?v=VgcwcmokPeQ)

My initial idea is to use the screen capturing feature of HyperHDR to record the very screen of the Raspberry Pi 4B that is both running the full screen video and HyperHDR. This way, I can assign each button to a different GPIO input that runs a video in full screen, and use the screen recording data to drive the LEDs.

I hope my explanation was thorough enough. Now to the questions:

1. Am I taking the right approach to this project?
2. Am I missing any hardware or software?
3. Is my RPI4B capable enough to smoothly and reliably run the script, the video and HyperHDR at the same time?
4. Has anybody done something similar to this? If so, then can you please point me in the right direction code-wise? As I said, I'm very new to this.

Thanks in advance to anyone who reads this and wants to help.

Cheers!
Juan

**Comment by awawa-dev** [comment]
Hi
> a strip of 10m of WS2811 addressable LEDs

Controlled directly from RPi? Bad idea, but it may work if you are lucky.

> My initial idea is to use the screen capturing feature of HyperHDR to record the very screen of the Raspberry Pi 4B that is both running the full screen video and HyperHDR. This way, I can assign each button to a different GPIO input that runs a video in full screen, and use the screen recording data to drive the LEDs.

Instead you can redirect LED colors output to the file to capture them (LED driver->Debug). Then you can either write a simple application to send them to HyperHDR in binary, or there is an even simpler solution: using a regular editor (preferably with the option to edit columns, such as Visual Code), at the beginning of each line add commands from the JSON API that set the selected previously saved color. You also need to add a delay at the end of each line. And you can probably run it even with a regular shell script. Capturing color to a file definitely works and is simpler than screen capturing , but I have not tested building JSON API commands (you have them in the advanced tab)

**Reply by mamanchekk** [reply]
Thanks for your quick reply :D

Is the LED strip too long for the controller to feed it data? Should I consider buying an ESP or something similar to drive all of the processing? I want this to work reliably and be able to replicate it many times when the prototyping phase has ended.

Regarding the second paragraph, I'm at a loss since I'm a bit of a newbie. I understand extracting color data from files is much more efficient than screen recording and I was able to finde the "file" controller type under the LED Hardware section. However, I'm afraid I didn't make myself fully understood:

What I want to do is use HyperHDR's capability of analyzing images (usually from a video capture card) and transforming them into LED light information, akin to how Pixel Mapping works. I have three video files with different animations that I want translated as pixel data for those LED strips.

What I've achieved so far is a simple python script that is controlled with three keyboard keys, where each keyboard keys switches to one of the three animations I've mentioned so far (I'll include reference images) that play full screen at the top part of the screen. 

<img width="1024" alt="Screenshot 2024-02-20 at 15 37 08" src="https://github.com/awawa-dev/HyperHDR/assets/116759637/67c9e044-d677-4d73-afae-810a2ea7526c">

After that, I wanted to use only that top portion of the screen to feed it via screen capture to hyper HDR so it can use that image to drive the LED strip lights like in the animations.

I'm aware this is a very inefficient way to do it but, honestly, my coding skills are not up there to fully understand what I should do better haha. Is there any way to feed the files directly to HyperHDR so that they can control the LED lights?

I'm very thankful for you support and your help. Much much much appreciated!

**Reply by awawa-dev** [reply]
> Is the LED strip too long for the controller to feed it data? Should I consider buying an ESP or something similar to drive all of the processing? I want this to work reliably and be able to replicate it many times when the prototyping phase has ended.

The SPI LEDs should work as it is a native protocol. Neopixels, such as the WS2811, require protocol emulation while maintaining tight timings, which is a difficult task for a real-time operating system on the RPi that has to deal with other things at the same time. The more LEDs, the greater the risk of problems and artifacts. Therefore, it is best to use esp or, even better, RPi Pico if we can use a cable connection. Regardless, the LEDs operate on 5V logic and the Raspberry Pi/ESP/Pico on 3.3V, so it is out of tolerance, although it may work sometimes. So you need a level shifter 3.3V to 5V for stable operation or the selected Pico/rp2040 model with a built-in level shifter. For these two reasons, I wrote that it may work with RPi and ws2811, if you are "lucky" (a combination of several favorable factors in unfavorable conditions that cannot be predicted as they out of the official spec)

> After that, I wanted to use only that top portion of the screen to feed it via screen capture to hyper HDR so it can use that image to drive the LED strip lights like in the animations.

And you will always use this top area to control the LEDs and switch between these 3 selected and previously recorded animations? Then you can still use the JSON API with recorded colors rather than images. In the photo you have only one LED configured in HyperHDR, but in the upper area you can switch it to many smaller ones so that each LED in the 10 meter strip (600?) is controlled from this upper area. Then play the animations once and capture LED colors.
